# Finding Customer Insights

---
 Needed to answer or explore things as below :

*   What is the overall customer response?
*   Which group of customer are needed to extra offer?
*   Predicting customers and their personality



# Data wrangling



*   Importing data set from my github repository
*   Explore Data and description
*   Identify and handle missing values
*   Data Formating
*   Data Standardization & normalization







In [106]:
import pandas as pd
import numpy as np
data = pd.read_csv('https://github.com/mdimran1/Dataset/blob/main/marketing_campaign.csv?raw=true', sep='\t') 


## Explore data and feature description

In [107]:
# data sample with both top & last 5 records
data.head(-5)

ID  Year_Birth   Education  ... Z_CostContact  Z_Revenue  Response
0     5524        1957  Graduation  ...             3         11         1
1     2174        1954  Graduation  ...             3         11         0
2     4141        1965  Graduation  ...             3         11         0
3     6182        1984  Graduation  ...             3         11         0
4     5324        1981         PhD  ...             3         11         0
...    ...         ...         ...  ...           ...        ...       ...
2230  7004        1984  Graduation  ...             3         11         0
2231  9817        1970      Master  ...             3         11         0
2232  8080        1986  Graduation  ...             3         11         0
2233  9432        1977  Graduation  ...             3         11         0
2234  8372        1974  Graduation  ...             3         11         0

[2235 rows x 29 columns]

In [108]:
# total rows and columns
print(f'Total rows : {data.shape[0]}')
print(f'Total columns : {data.shape[1]}') 

Total rows : 2240
Total columns : 29


In [109]:
# data information with data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [110]:
data.describe()

ID   Year_Birth  ...  Z_Revenue     Response
count   2240.000000  2240.000000  ...     2240.0  2240.000000
mean    5592.159821  1968.805804  ...       11.0     0.149107
std     3246.662198    11.984069  ...        0.0     0.356274
min        0.000000  1893.000000  ...       11.0     0.000000
25%     2828.250000  1959.000000  ...       11.0     0.000000
50%     5458.500000  1970.000000  ...       11.0     0.000000
75%     8427.750000  1977.000000  ...       11.0     0.000000
max    11191.000000  1996.000000  ...       11.0     1.000000

[8 rows x 26 columns]

In [111]:
data.describe(include=object)

Education Marital_Status Dt_Customer
count         2240           2240        2240
unique           5              8         663
top     Graduation        Married  31-08-2012
freq          1127            864          12

In [112]:
# explore column values
for i in data.columns:
  print(i)
  print (data[i].value_counts())
  print()


ID
2546     1
967      1
5396     1
5394     1
7441     1
        ..
5654     1
2698     1
11133    1
3389     1
0        1
Name: ID, Length: 2240, dtype: int64

Year_Birth
1976    89
1971    87
1975    83
1972    79
1978    77
1970    77
1973    74
1965    74
1969    71
1974    69
1956    55
1958    53
1979    53
1977    52
1952    52
1968    51
1959    51
1954    50
1966    50
1960    49
1955    49
1982    45
1963    45
1967    44
1962    44
1951    43
1957    43
1964    42
1983    42
1986    42
1980    39
1981    39
1984    38
1961    36
1953    35
1985    32
1989    30
1949    30
1988    29
1950    29
1987    27
1948    21
1990    18
1946    16
1947    16
1991    15
1992    13
1945     8
1944     7
1943     7
1995     5
1993     5
1994     3
1996     2
1940     1
1893     1
1899     1
1941     1
1900     1
Name: Year_Birth, dtype: int64

Education
Graduation    1127
PhD            486
Master         370
2n Cycle       203
Basic           54
Name: Education, dtype: int64

Marital_St

<b>Dataset contains total 2240 records and 29 features. Target column or feature is response which indicates binary records of the latest promotion that customers has been taken. Primary Feature decription are as below :</b>



0   ID : Customer's unique identity

1   Year_Birth : Customer's birth year & easy to find age

2   Education : Maximum education level

3   Marital_Status : married or others

4   Income : Amount of earning, Needed to normalize or binning 

5   Kidhome : No. of kids [ according google teen ages = under 12]

6   Teenhome : No. of teens [ according google teen ages = 13 to 19]

7   Dt_Customer : When customers are envolved

8   Recency : count of days last ordered

9   MntWines : Total amount as spent on wine

10   MntFruits : Total amount as spent on Fruits

11   MntMeatProducts : Total amount as spent on Meat

12   MntFishProducts : Total amount as spent on Fish

13   MntSweetProducts : Total amount as spent on Sweet

14   MntGoldProds : Total amount as spent on Gold

15   NumDealsPurchases : Counts of purchases with discount

16   NumWebPurchases : purchased by company’s web site

17   NumCatalogPurchases : Number of purchases made using a catalogue

18   NumStorePurchases : Number of purchases from store

19   NumWebVisitsMonth : Number of visits to company’s web site 

20   AcceptedCmp3 : Does customer accepted the offer campaign 3?

21   AcceptedCmp4 : Does customer accepted the offer campaign 4?

22   AcceptedCmp5 : Does customer accepted the offer campaign 5?

23   AcceptedCmp1 : Does customer accepted the offer campaign 1?

24   AcceptedCmp2 : Does customer accepted the offer campaign 2?

25   Complain : Has any cpmplain? (1 or 0)

26   Z_CostContact : This feature has constant 3 only

27   Z_Revenue : This feature has constant 11 only

28   Response: Target feature

## Identify and handle missing values

---
Based on summary above every column has total 2240 rows except Income column.<br>Income column has 24 Nan values.<br>Income column is about to customer description or profile details. Its actually a independent feature. <br> It is better to <b> Replace by Mean</b> as 52247.25


In [113]:
data.isnull().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

In [114]:
# income column showing total 24 records that are null 

# Looking top 5 null records
data[data.Income.isnull()].head()

# we need to replace (with mean, freq or other function) or drop these Nan value

ID  Year_Birth   Education  ... Z_CostContact  Z_Revenue  Response
10  1994        1983  Graduation  ...             3         11         0
27  5255        1986  Graduation  ...             3         11         0
43  7281        1959         PhD  ...             3         11         0
48  7244        1951  Graduation  ...             3         11         0
58  8557        1982  Graduation  ...             3         11         0

[5 rows x 29 columns]

In [115]:
data.Income.describe()

count      2216.000000
mean      52247.251354
std       25173.076661
min        1730.000000
25%       35303.000000
50%       51381.500000
75%       68522.000000
max      666666.000000
Name: Income, dtype: float64

In [116]:
data.Income.value_counts()#.count()

7500.0     12
35860.0     4
18929.0     3
34176.0     3
48432.0     3
           ..
71022.0     1
30675.0     1
30168.0     1
19656.0     1
41728.0     1
Name: Income, Length: 1974, dtype: int64

In [117]:
# Mean of Income:
avg_income = data['Income'].mean()

# replace nan by mean
data['Income'].fillna(avg_income, inplace=True)

## Data Duplicacy Removing

---

Here has no duplicate records or row. Data is free from any kinds of duplicacy.

In [118]:
# Checking duplicacy
data.duplicated().sum()
data.duplicated(subset=['ID','Income' ]).sum()

0

## Data formating 

---
<b>Task Summery:</b> 
<br>1. There are 3 'object type' variables as named: 'Education', 'Marital_Status', 'Dt_Customer' and the rest are all numerical format.
<br>2. 'Dt_customer' data is in object format but its should be in Datetime formation.
<br>3. Others Formation are still good.


In [119]:
#Transform items of dt_customer from obj to datetime foramat 
data['Dt_Customer']= pd.to_datetime(data['Dt_Customer'])
#data.Dt_Customer

## Data Standardization & normalization

---
<b>Task Summery: </b>
<br>1. Year Birth to transform age is more informative and actionable.
<br>2. Marital_Status has 8 unique values. But <b> Married, Single, Divorced</b> these 3 are suitable for this project.
<br>3. Second cycle corresponds to graduate level or master's level studies. Third cycle corresponds to doctoral or PhD level studies. So we will replace education column with Graduation for masters and 2nd cycle.
<br>4. Creating a column as Total_spends. As adding all the amounts spend in different products or services is very effective to analysis.


In [120]:
# assigning new column as Age from Year birthdate
data['Age'] = 2021 - data['Year_Birth']

In [121]:
# Finding useless columns
data.nunique().sort_values()

Z_Revenue                 1
Z_CostContact             1
AcceptedCmp2              2
AcceptedCmp4              2
AcceptedCmp3              2
Response                  2
AcceptedCmp5              2
Complain                  2
AcceptedCmp1              2
Teenhome                  3
Kidhome                   3
Education                 5
Marital_Status            8
NumCatalogPurchases      14
NumStorePurchases        14
NumDealsPurchases        15
NumWebPurchases          15
NumWebVisitsMonth        16
Year_Birth               59
Age                      59
Recency                 100
MntFruits               158
MntSweetProducts        177
MntFishProducts         182
MntGoldProds            213
MntMeatProducts         558
Dt_Customer             663
MntWines                776
Income                 1975
ID                     2240
dtype: int64

In [122]:
# Transforming marital_status column
data['Marital_Status'].replace('Together', 'Married', inplace=True )
data['Marital_Status'].replace('Widow', 'Divorced', inplace=True )
data['Marital_Status'].replace( ['Alone','YOLO', 'Absurd'], 'Single', inplace=True )

In [123]:
# education column tranformation:
data['Education'].replace(['Master', '2n Cycle'	],'Graduation', inplace=True )

#checking Education column
data.Education.value_counts().to_frame()

Education
Graduation       1700
PhD               486
Basic              54

In [124]:
# Forming 'Total_spends' column

data['Total_spends']= data.loc[:, 'MntWines':'MntGoldProds'].sum(axis=1)

In [125]:
# exploring Total_spends column

# before that Changing column index:
cols = data.columns.tolist()
cols = cols[0:15] + cols[30:] +cols[15:30]
data = data[cols]

# exploring Total_spends column
data.loc[:,'MntWines':'Total_spends']

MntWines  MntFruits  ...  MntGoldProds  Total_spends
0          635         88  ...            88          1617
1           11          1  ...             6            27
2          426         49  ...            42           776
3           11          4  ...             5            53
4          173         43  ...            15           422
...        ...        ...  ...           ...           ...
2235       709         43  ...           247          1341
2236       406          0  ...             8           444
2237       908         48  ...            24          1241
2238       428         30  ...            61           843
2239        84          3  ...            21           172

[2240 rows x 7 columns]

# New Section